media/src/commonMain/kotlin/MediaManager.kt

```diff
@@ -0,0 +1,266 @@
+
+    private val mediaMutex = Mutex()
+    private var mediaSurfaceJob: Job? = null
+
+    private val seekMutex = Mutex()
+    private var activeSeek: Pair<Duration, CompletableDeferred<Boolean>>? = null
+    private var queuedSeek: Pair<Duration, CompletableDeferred<Boolean>>? = null
+
+    final override suspend fun createPlayableMedia(source: MediaSource): PlayableMedia? = mediaMutex.withLock {
+        handleCreatePlayableMedia(source).also {
+            mediaSurfaceJob?.cancelAndJoin()
+            mediaSurfaceJob = mediaSurfaceProvider?.let {
+                this@BaseMediaManager.launch {
+                    mediaSurfaceProvider.surface.onCompletion {
+                        renderVideoOnSurface(null)
+                    }.collect {
+                        renderVideoOnSurface(it)
+                    }
+                }
+            }
+        }
+    }
+
+    protected abstract fun handleCreatePlayableMedia(source: MediaSource): PlayableMedia?
+
+    protected open fun handlePrepared(playableMedia: PlayableMedia) {
+        _events.trySend(MediaManager.Event.DidPrepare(playableMedia))
+    }
+
+    protected open fun handleError(error: PlaybackError) {
+        _events.trySend(MediaManager.Event.DidFailWithError(error))
+    }
+
+    protected open fun handleCompleted() {
+        _events.trySend(MediaManager.Event.DidComplete)
+    }
+
+    final override suspend fun seekTo(duration: Duration): Boolean {
+        val result = CompletableDeferred<Boolean>()
+        return seekMutex.withLock {
+            val queuedSeek = queuedSeek
+            when {
+                activeSeek == null -> {
+                    activeSeek = duration to result
+                    startSeek(duration)
+                    result
+                }
+                queuedSeek != null && queuedSeek.first == duration -> queuedSeek.second
+                else -> {
+                    this.queuedSeek?.second?.complete(false)
+                    this.queuedSeek = duration to result
+                    result
+                }
+            }
+        }.await()
+    }
+
+    final override fun reset() {
+        mediaSurfaceJob?.cancel()
+        handleReset()
+    }
+
+    protected abstract fun handleReset()
+
+    override fun end() {
+        cleanUp()
+        _events.trySend(MediaManager.Event.DidEnd)
+    }
+
+    protected open fun handleSeekCompleted(success: Boolean) {
+        launch {
+            seekMutex.withLock {
+                activeSeek?.second?.complete(success)
+                val queuedSeek = queuedSeek
+                activeSeek = when {
+                    queuedSeek == null -> null
+                    activeSeek?.first == queuedSeek.first -> {
+                        queuedSeek.second.complete(success)
+                        null
+                    }
+                    else -> queuedSeek
+                }
+                activeSeek?.let { startSeek(it.first) }
```

Comment:
It seems maybe having Seekable in the main state hierarchy would also work.

Reply:
The mechanism for seeking like this is pretty nifty btw (active and a single queued item with callbacks).

Reply:
I didnt make it part of the state machine cause Android doesnt consider it either. Doesn't make it impossible to make this of course, but it does complicate things since seeking can occur while playing/pauses/completed and the behaviour of the internal platforms while seeking is somewhat unclear.